File: java/515.FindLargestValueinEachTreeRow.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
 
public class Solution {
    private static final int FIRST_LEVEL = 0;
    
    
    private List<Integer> extendArray(List<Integer> in, int size){
        while(in.size() < size){
            in.add(Integer.MIN_VALUE);
        }
        return in;
    }
    
    public List<Integer> largestValues(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        
        Deque<TreeNodeProxy> stack = new ArrayDeque<TreeNodeProxy>();
        stack.push(new TreeNodeProxy(root, FIRST_LEVEL));
        
        while(!stack.isEmpty()){
            TreeNodeProxy tnp = stack.pop();
            TreeNode node = tnp.node;
            if (node == null){
                continue;
            }
            
            int level = tnp.level;
            int value = node.val;
            
            if (ret.size() < level + 1){
                ret = extendArray(ret, level + 1);
            }
            
            if (ret.get(level) < value){
                ret.set(level, value);
            }
            
            if (node.left != null) stack.push(new TreeNodeProxy(node.left, level + 1));
            if (node.right != null) stack.push(new TreeNodeProxy(node.right, level + 1));
        }
        
        return ret;
        
    }
}



class TreeNodeProxy {
    
    int level;
    TreeNode node;
    
    TreeNodeProxy(TreeNode node, int level){
        this.node = node;
        this.level = level;
    }
}

File: swift/515.FindLargestValueinEachTreeRow.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
#########BFS   69ms #########
class Solution {
     

    func largestValues(_ root: TreeNode?) -> [Int] {
        var maxes = [Int]()  // 创建一个数组，声明+初始化
        var row = [root]
        
        while row.count > 0 {
            var rowValue = [Int]()
            for r in row {
                if let node = r{
                    rowValue.append(node.val)
                }
            }
            if let m = rowValue.max(){  
                maxes.append(m)
            }else{  // 如果连最大值都没有，那只能说明这一层都是nil，break掉
                break
            }
            
            var nextRow = [TreeNode?]()
            for r in row {
                if let node = r {
                    nextRow.append(node.left)
                    nextRow.append(node.right)
                }
            }
            
            row = nextRow
            
        }
        
        return maxes
    }

}


 #######DFS  72ms##########
class Solution {
    var maxes = [Int]()  // 创建一个数组，声明+初始化 

    func largestValues(_ root: TreeNode?) -> [Int] {
          // 没有；
        findLargest(root, 0)
        
        return maxes
    }
    
    // 参数传递是这样写的  level：Int
    func findLargest(_ root: TreeNode?, _ level: Int){
        if let node = root { // unwrap
            var value = node.val
            if maxes.count == level{
                maxes.append(value)
            }else{
                maxes[level] = max(value, maxes[level])
            }
            
            findLargest(node.left, level + 1)
            findLargest(node.right, level + 1)
        }
    }
}